lib/actions/build.coffee

```diff
@@ -0,0 +1,224 @@
+Promise = require('bluebird')
+tar = require('tar-stream')
+dockerBuild = require('resin-docker-build')
+resolve = require('resin-bundle-resolve')
+klaw = require('klaw')
+path = require('path')
+fs = require('mz/fs')
+_ = require('lodash')
+
+# When we do TLS enabled docker builds, node sometimes complains
+# about the certificates not being fully authorized, this stops
+# that.
+process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0'
+
+tarDirectory = (dir) ->
+	getFiles = (dir) ->
+		new Promise (resolve, reject) ->
+			files = []
+			klaw(dir)
+			.on 'data', (item) ->
+				files.push(item.path) if not item.stats.isDirectory()
+			.on 'end', ->
+				resolve(files)
+			.on('error', reject)
+
+	pack = tar.pack()
+	getFiles(dir)
+	.map (file) ->
+		relPath = path.relative(path.resolve(dir), file)
+		Promise.all([relPath, fs.stat(file), fs.readFile(file)])
+	.map ([filename, stats, data]) ->
+		pack.entryAsync({ name: filename, size: stats.size }, data)
+	.then ->
+		pack.finalize()
+		return pack
+
+getAppInfo = (application) ->
+	resin = require('resin-sdk-preconfigured')
+	return resin.models.application.get(application)
+	.then (app) ->
+		if !app?
+			throw new Error('Could not read application information!')
+		Promise.all([app, resin.models.config.getDeviceTypes()])
+	.then ([app, config]) ->
+		config = _.find(config, 'slug': app.device_type)
+		if !config?
+			throw new Error('Could not read application information!')
+		app.arch = config.arch
+		return app
+
+getBundleInfo = (options) ->
+	if options.application?
+		# An application was provided
+		getAppInfo(options.application)
+		.then (app) ->
+			console.log('Building for application: ' + app.app_name)
+			return [app.arch, app.device_type]
+		.catch (err) ->
+			throw err
+	else if options.arch? and options.deviceType?
+		return Promise.resolve([options.arch, options.deviceType])
+	else
+		# No information, cannot do resolution
+		return
+
+generateConnectOpts = (options) ->
+	connectOpts = {}
+	# Firsly need to decide between a local docker socket
+	# and a host available over a host:port combo
+	if options.docker? and not options.dockerHost?
+		# good, local docker socket
+		connectOpts['socketPath'] = options.docker
+	else if options.dockerHost? and not options.docker?
+		# Good a host is provided, and local socket isn't
+		connectOpts['host'] = options.dockerHost
+		connectOpts['port'] = options.dockerPort || 2376
+	else if options.docker? and options.dockerHost?
+		# Both provided, no obvious way to continue
+		throw new Error("Both a local docker socket and docker host have been provided. Don't know how to continue.")
+	else
+		# None provided, assume default docker local socket
+		connectOpts['socketPath'] = '/var/run/docker.sock'
+
+	# Now need to check if the user wants to connect over TLS
+	# to the host
+	connectOpts['ca'] = options.ca if options.ca?
+	connectOpts['cert'] = options.cert if options.cert?
+	connectOpts['key'] = options.key if options.key?
```

Comment:
I think if you don't provide all three of these, Dockerode will [silently not use TLS](https://github.com/apocas/docker-modem/blob/master/lib/modem.js#L68). We should probably throw an error to force you to provide all three or none.